ingesters/SimpleRelay/json.go

```diff
@@ -217,65 +231,106 @@ func jsonConnHandler(c net.Conn, cfg jsonHandlerConfig, igst *ingest.IngestMuxer
 		}
 		tg, err = timegrinder.NewTimeGrinder(tcfg)
 		if err != nil {
-			lg.Error("failed to get a handle on the timegrinder", log.KVErr(err))
+			ll.Error("failed to get a handle on the timegrinder", log.KVErr(err))
 			return
 		} else if err = cfg.timeFormats.LoadFormats(tg); err != nil {
-			lg.Error("failed to load custom time formats", log.KVErr(err))
+			ll.Error("failed to load custom time formats", log.KVErr(err))
 			return
 		}
 		if cfg.setLocalTime {
 			tg.SetLocalTime()
 		}
 		if cfg.timezoneOverride != `` {
-			err = tg.SetTimezone(cfg.timezoneOverride)
-			if err != nil {
-				lg.Error("failed to set timezone", log.KV("timezone", cfg.timezoneOverride), log.KVErr(err))
+			if err = tg.SetTimezone(cfg.timezoneOverride); err != nil {
+				ll.Error("failed to set timezone", log.KV("timezone", cfg.timezoneOverride), log.KVErr(err))
 				return
 			}
 		}
 		if cfg.formatOverride != `` {
 			if err = tg.SetFormatOverride(cfg.formatOverride); err != nil {
-				lg.Error("Failed to load format override", log.KV("override", cfg.formatOverride), log.KVErr(err))
+				ll.Error("Failed to load format override", log.KV("override", cfg.formatOverride), log.KVErr(err))
 				return
 			}
 		}
 	}
-	bio := bufio.NewReader(c)
+
+	lr := &io.LimitedReader{
+		R: bufio.NewReader(c),
+		N: cfg.maxObjectSize,
+	}
+	dec := json.NewDecoder(lr)
 	for {
-		//get the data entry and clean it a bit
-		data, err := bio.ReadBytes('\n')
-		if err != nil {
-			break
+		var obj json.RawMessage
+		if err := dec.Decode(&obj); err != nil {
+			// check if limited reader is exhausted so that we can throw a better error
+			// we do not log a standard EOF, will log if EOF and N==0 because that means
+			// we got an oversized object and want to honk about it.  The json decoder
+			// will also likely honk with something other than pure EOF if it gets an
+			// EOF while decoding an object, so we honk about that too
+			if lr.N == 0 {
+				if err == io.EOF || errors.Is(err, io.EOF) {
+					ll.Error("oversized json object", log.KV("max-size", cfg.maxObjectSize))
+				} else {
+					ll.Error("oversized json object", log.KV("max-size", cfg.maxObjectSize), log.KVErr(err))
+				}
+			} else if err != io.EOF {
+				//just a plain old error
+				ll.Error("invalid json object", log.KV("max-size", cfg.maxObjectSize), log.KVErr(err))
+			}
+			break // we pretty much have to just hang up
```

Comment:
this has the same effect as calling return, right? I'm leery of break, because for loops and other logic have a tendency to multiply, and before long break is *not* the same as return.

Reply:
it does, but good point, will address.